File: rubygit/git.rb
require "fileutils"
require "pathname"

require_relative "./database"
require_relative "./workspace"

command = ARGV.shift

case command
when "init"
  path = ARGV.fetch(0, Dir.getwd)
  root_path = Pathname.new(File.expand_path(path))
  git_path = root_path.join(".git")

  ["objects", "refs"].each do |dir|
    begin
      FileUtils.mkdir_p(git_path.join(dir))
    rescue Errno::EACCES => error
      $stderr.puts "fatal: #{ error.message }"
      exit 1
    end
  end

  puts "Initialized empty git repository in #{ git_path }"
  exit 0
when "commit"
  root_path = Pathname.new(Dir.getwd)
  git_path = root_path.join(".git")
  db_path = git_path.join("objects")
  workspace = Workspace.new(root_path)
  database = Database.new(db_path)
  workspace.list_files.each do |path|
    data = workspace.read_file(path)
    blob = Blob.new(data)
    database.store(blob)
  end
else
  $stderr.puts "git: '#{ command }' is not a git command."
  exit 1
end

File: pygit/git.py
"""Git command line interface."""
# pylint: disable=C0103
import os
import sys
import pathlib


command = sys.argv[1]

if command == 'init':
    path = os.path.abspath(sys.argv[2])
    root_path = pathlib.Path(path)
    git_path = root_path.joinpath('.git')

    for dirname in ['objects', 'refs']:
        try:
            git_path.joinpath(dirname).mkdir(parents=True, exist_ok=False)
        except Exception as err:
            print(f'Fatal: {err}', file=sys.stderr)
            sys.exit(1)

    print(f'Initialized empty git repository in {git_path}')
elif command == 'commit':
    from blob import Blob
    from database import Database
    from workspace import Workspace
    root_path = os.getcwd()
    db_path = os.path.join(root_path, '.git', 'objects')

    workspace = Workspace(os.getcwd())
    database = Database(db_path)
    for name in workspace.list_files():
        data = workspace.read_file(name)
        blob = Blob(data)
        database.store(blob)
else:
    print(f'git: {command!r} is not a git command.', file=sys.stderr)
    sys.exit(1)
